### PR TITLE
[KYUUBI #1962] Add timeout check for createSpark

### LINKED ISSUE #1962: [Improvement] Make sparkSession or SparkContext as a scala.concurrent.Future to give it proper timeout
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

https://github.com/apache/incubator-kyuubi/issues/1960 will fix a certain problem with master deploy mode, for client mode, the timeout describe there will almost not happen. The sc might get stuck will the driver is on and wait for the ApplicaitonMaster to be allocated. Although spark will have a timeout for checking for driver-am connected. But this may not happen when the AM is not able to be allocated in time.

### How should we improve?

We need a timeout for this whole progress.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

close #1962

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [X] Add screenshots for manual tests if appropriate
image

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by wForget** [comment]
I reimplemented #1962 based on [#issuecomment-1066329556](https://github.com/apache/incubator-kyuubi/pull/1998#issuecomment-1066329556). @ulysses-you  could you help me review this?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 1]
```diff
@@ -195,8 +196,11 @@ object SparkSQLEngine extends Logging {
         s" and submitted at $submitTime.")
     } else {
       var spark: SparkSession = null
+      val checkFlag = new AtomicBoolean(false)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:199]: we can define it as global, and the name should be refine, maybe like `sparkSessionCreated`

Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:199]: > we can define it as global, and the name should be refine, maybe like `sparkSessionCreated`

thanks, i will do that

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/IndividualSparkSuite.scala [hunk 2]
```diff
@@ -80,6 +82,47 @@ class SparkEngineSuites extends KyuubiFunSuite {
...
+      val e1 = intercept[KyuubiException] {
+        SparkSQLEngine.main(Array.empty)
+      }.getMessage
+      assert(SparkSQLEngine.currentEngine.isEmpty)
+      assert(e1.startsWith("The total engine initialization time"))
+    }
+  }
+
+  test("test engine `createSpark` timeout") {
+    val timeout = 10000
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/IndividualSparkSuite.scala:102]: we can reduce the time, 3s is enough ?

**Review by ulysses-you — APPROVED** [review]

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/IndividualSparkSuite.scala (hunk 2 above)
Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/IndividualSparkSuite.scala]: > we can reduce the time, 3s is enough ?

done

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**